### ISSUE TITLE
Optional argument taking multiline closure is indented strangely

### ISSUE THREAD

**Issue by nex3** [dart-lang/dart_style#568]
The following code:

```dart
void main() {
  var controller = new StreamController<T>(sync: true, onCancel: () {
    return Future.wait(operationSet.map((operation) => operation.cancel()));
  });
}
```

is formatted as:

```dart
void main() {
  var controller = new StreamController<T>(
      sync: true,
      onCancel: () {
        return Future.wait(operationSet.map((operation) => operation.cancel()));
      });
}
```

This seems substantially worse than the input. Note that it has similar formatting no matter how long the `onCancel` callback is, which can lead to a lot of unnecessary indentation.

**Comment by munificent** [comment]
Hmm, this one is surprising. I would expect it to keep your original formatting.

**Comment by munificent** [comment]
This one is a little tricky. The formatter has special rules to handle functions in argument lists and make them more block-like, for cases like:

```dart
method(thing: () {
  ...
}, other: () {
  ...
});
```

But in some cases applying that style would produce weird output:

```dart
method(leadingNamed: arg, 
    another: arg,
    thing: () {
  ...
}, other: () {
  ...
},
    trailingNamed: arg);
```

Better would be:

```dart
method(leadingNamed: arg, 
    another: arg,
    thing: () {
      ...
    }, other: () {
      ...
    },
    trailingNamed: arg);
```

Ideally, it would decide whether to indent function bodies in cases like this based on whether or not the named arguments needed to split. Unfortunately, that's really hard with how the formatter's internal representation works.

Instead, it has some rules to try to avoid nasty cases like this. In particular, one rule is "if there are named arguments and at least one of them is *not* a function, don't do special block-like indentation for functions". That's what you're running into here.

I've tried tweaking that rule a few ways, but everything I tried made more code worse than it made better. Leaving this bug open for now.
